File: src/leetCodeProblems/Sorting/ArrayMatrixTwoD/MaxUnitsOnTruck1710.go
package main

import (
	"log"
	"sort"
)

/*
- LeetCode - https://leetcode.com/problems/maximum-units-on-a-truck/
- Time - O(nlogn)
- Space - O(1)
*/

func maximumUnits(boxTypes [][]int, truckSize int) int {
	sort.SliceStable(boxTypes, func(i, j int) bool {
		return boxTypes[i][1] > boxTypes[j][1]
	})
	out := 0

	for _, v := range boxTypes {
		if v[0] <= truckSize {
			out += v[0] * v[1]
			truckSize -= v[0]
		} else {
			out += truckSize * v[1]
			truckSize = 0
			break
		}
	}
	return out
}

func main() {
	log.Println(maximumUnits([][]int{{1, 3}, {2, 2}, {3, 1}}, 4))
	log.Println(maximumUnits([][]int{{5, 10}, {2, 5}, {4, 7}, {3, 9}}, 10))
}

File: src/leetCodeProblems/Sorting/ArrayMatrixTwoD/MaxUnitsOnTruck1710.java
package leetCodeProblems.Sorting;

/**
 * LeetCode - https://leetcode.com/problems/maximum-units-on-a-truck/solution/
 *
 * TimeComplexity - O(nlogn)
 * SpaceComplexity - O(1)
 */

import java.util.Arrays;
import java.util.Comparator;

public class MaxUnitsOnTruck1710 {

    static class UnitsComparator implements Comparator<int[]> {

        public int compare(int[] a, int[] b) {
            return b[1]-a[1];
        }
    }
    public int maximumUnits(int[][] boxTypes, int truckSize) {

        Arrays.sort(boxTypes, new UnitsComparator());

        //System.out.println(Arrays.toString(boxTypes));

        int remainingSizeCounter = truckSize;
        int answer = 0;

        for (int i=0; i < boxTypes.length; i++) {

            //System.out.println(Arrays.toString(boxTypes[i]));

            if (boxTypes[i][0] <= remainingSizeCounter) {
                answer += boxTypes[i][0]*boxTypes[i][1];
                remainingSizeCounter -= boxTypes[i][0];
            }
            else {
                answer += remainingSizeCounter*boxTypes[i][1];
                remainingSizeCounter = 0;
            }
        }

        return answer;
    }

    public static void main(String[] args) {

        int[][] boxes = {{1, 3}, {2, 2}, {3,1}};
        int truckSize = 4;

        MaxUnitsOnTruck1710 obj = new MaxUnitsOnTruck1710();
        System.out.println(obj.maximumUnits(boxes, truckSize));
    }
}
